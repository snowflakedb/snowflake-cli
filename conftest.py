import sys
from unittest import mock

import pytest
from prompt_toolkit.output import DummyOutput
from rich.panel import Panel


@pytest.fixture(name="win32_dummy_console", autouse=True)
def make_win32_dummy_console():
    """Windows in CI/CD does not provide full terminal.
    We need to patch default output with DummyOutput()
    DummyOutput does not provide detection of the capabilities
    and we need to force rich Panel not to play safe.
    """
    if sys.platform == "win32":
        to_patch = "prompt_toolkit.output.defaults.create_output"
        with mock.patch(to_patch, return_value=DummyOutput()):
            with mock.patch("snowflake.cli.api.console.console.Panel") as mock_panel:
                mock_panel.side_effect = lambda *args, **kwargs: Panel(
                    *args,
                    **kwargs,
                    safe_box=False,
                )
                yield
    else:
        yield
