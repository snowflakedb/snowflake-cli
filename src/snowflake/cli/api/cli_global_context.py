# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field, replace
from functools import wraps
from pathlib import Path
from typing import TYPE_CHECKING, Iterator, Optional

from snowflake.cli.api.connections import ConnectionContext, OpenConnectionCache
from snowflake.cli.api.exceptions import MissingConfiguration
from snowflake.cli.api.metrics import CLIMetrics
from snowflake.cli.api.output.formats import OutputFormat
from snowflake.cli.api.rendering.jinja import CONTEXT_KEY
from snowflake.connector import SnowflakeConnection

if TYPE_CHECKING:
    from snowflake.cli.api.project.definition_manager import DefinitionManager
    from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition

_CONNECTION_CACHE = OpenConnectionCache()


@dataclass
class _CliGlobalContextManager:
    connection_context: ConnectionContext = field(default_factory=ConnectionContext)
    connection_cache: OpenConnectionCache = (
        _CONNECTION_CACHE  # by default, use global cache
    )

    output_format: OutputFormat = OutputFormat.TABLE
    silent: bool = False
    verbose: bool = False
    experimental: bool = False
    enable_tracebacks: bool = True

    metrics: CLIMetrics = field(default_factory=CLIMetrics)

    project_path_arg: str | None = None
    project_is_optional: bool = True
    project_env_overrides_args: dict[str, str] = field(default_factory=dict)

    # FIXME: this property only exists to help implement
    # nativeapp_definition_v2_to_v1 and single_app_and_package.
    # Consider changing the way this calculation is provided to commands
    # in order to remove this logic (then make project_definition a non-cloned @property)
    override_project_definition: ProjectDefinition | None = None

    _definition_manager: DefinitionManager | None = None

    # which properties invalidate our current DefinitionManager?
    DEFINITION_MANAGER_DEPENDENCIES = [
        "project_path_arg",
        "project_is_optional",
        "project_env_overrides_args",
    ]

    def reset(self):
        self.__init__()

    def clone(self) -> _CliGlobalContextManager:
        return replace(
            self,
            connection_context=self.connection_context.clone(),
            project_env_overrides_args=self.project_env_overrides_args.copy(),
            metrics=self.metrics.clone(),
        )

    def __setattr__(self, prop, val):
        if prop in self.DEFINITION_MANAGER_DEPENDENCIES:
            self._clear_definition_manager()

        super().__setattr__(prop, val)

    @property
    def project_definition(self) -> ProjectDefinition | None:
        if self.override_project_definition:
            return self.override_project_definition

        return self._definition_manager_or_raise().project_definition

    @property
    def project_root(self) -> Path:
        return Path(self._definition_manager_or_raise().project_root)

    @property
    def template_context(self) -> dict:
        return self._definition_manager_or_raise().template_context

    @property
    def connection(self) -> SnowflakeConnection:
        """
        Returns a connection for our configured context from the configured cache.
        By default, this is the global _CONNECTION_CACHE. If a matching connection
        does not already exist, creates a new connection and caches it.
        """
        self.connection_context.validate_and_complete()
        return self.connection_cache[self.connection_context]

    def _definition_manager_or_raise(self) -> DefinitionManager:
        """
        (Re-)parses project definition based on project args (project_path_arg and
        project_env_overrides_args). If we cannot provide a project definition
        (i.e. no snowflake.yml) and require one, raises MissingConfiguration.
        """
        from snowflake.cli.api.project.definition_manager import DefinitionManager

        # don't need to re-parse definition if we already have one
        if not self._definition_manager:
            dm = DefinitionManager(
                self.project_path_arg,
                {CONTEXT_KEY: {"env": self.project_env_overrides_args}},
            )
            if not dm.has_definition_file and not self.project_is_optional:
                raise MissingConfiguration(
                    "Cannot find project definition (snowflake.yml). Please provide a path to the project or run this command in a valid project directory."
                )
            self._definition_manager = dm

        return self._definition_manager

    def _clear_definition_manager(self):
        """
        Force re-calculation of definition_manager and its dependent attributes
        (template_context, project_definition, and project_root).
        """
        self._definition_manager = None


class _CliGlobalContextAccess:
    def __init__(self, manager: _CliGlobalContextManager):
        self._manager = manager

    @property
    def connection(self) -> SnowflakeConnection:
        return self._manager.connection

    @property
    def connection_context(self) -> ConnectionContext:
        return self._manager.connection_context

    @property
    def enable_tracebacks(self) -> bool:
        return self._manager.enable_tracebacks

    @property
    def metrics(self) -> CLIMetrics:
        return self._manager.metrics

    @property
    def output_format(self) -> OutputFormat:
        return self._manager.output_format

    @property
    def verbose(self) -> bool:
        return self._manager.verbose

    @property
    def experimental(self) -> bool:
        return self._manager.experimental

    @property
    def project_definition(self) -> ProjectDefinition | None:
        return self._manager.project_definition

    @property
    def project_root(self) -> Path | None:
        return self._manager.project_root

    @property
    def template_context(self) -> dict:
        return self._manager.template_context

    @property
    def silent(self) -> bool:
        if self._should_force_mute_intermediate_output:
            return True
        return self._manager.silent

    @property
    def _should_force_mute_intermediate_output(self) -> bool:
        """Computes whether cli_console output should be muted."""
        return self._manager.output_format == OutputFormat.JSON

    @property
    def snow_api_root(
        self,
    ) -> Optional[
        object
    ]:  # Should be Optional[Root], but we need local import for performance reasons
        from snowflake.core import Root

        if self.connection:
            return Root(self.connection)
        else:
            return None


_CLI_CONTEXT_MANAGER: ContextVar[_CliGlobalContextManager | None] = ContextVar(
    "cli_context", default=None
)


def get_cli_context_manager() -> _CliGlobalContextManager:
    mgr = _CLI_CONTEXT_MANAGER.get()
    if not mgr:
        mgr = _CliGlobalContextManager()
        _CLI_CONTEXT_MANAGER.set(mgr)
    return mgr


def get_cli_context() -> _CliGlobalContextAccess:
    return _CliGlobalContextAccess(get_cli_context_manager())


def span(span_name: str):
    """
    Decorator to start a command metrics span that encompasses a whole function

    Must be used instead of directly calling @get_cli_context().metrics.span(span_name)
    as a decorator to ensure that the cli context is grabbed at run time instead of at
    module load time, which would not reflect forking
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with get_cli_context().metrics.span(span_name):
                return func(*args, **kwargs)

        return wrapper

    return decorator


@contextmanager
def fork_cli_context(
    connection_overrides: dict | None = None,
    project_env_overrides: dict[str, str] | None = None,
    project_is_optional: bool | None = None,
    project_path: str | None = None,
) -> Iterator[_CliGlobalContextAccess]:
    """
    Forks the global CLI context, making changes that are only visible
    (e.g. via get_cli_context()) while inside this context manager.

    Please note that environment variable changes are only visible through
    the project definition; os.getenv / os.environ / get_env_value are not
    affected by these new values.
    """
    old_manager = get_cli_context_manager()
    new_manager = old_manager.clone()
    token = _CLI_CONTEXT_MANAGER.set(new_manager)

    if connection_overrides:
        new_manager.connection_context.update(**connection_overrides)

    if project_env_overrides:
        new_manager.project_env_overrides_args.update(project_env_overrides)

    if project_is_optional is not None:
        new_manager.project_is_optional = project_is_optional

    if project_path:
        new_manager.project_path_arg = project_path

    yield _CliGlobalContextAccess(new_manager)
    _CLI_CONTEXT_MANAGER.reset(token)
