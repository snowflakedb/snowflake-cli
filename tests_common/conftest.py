# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import importlib
import os
import shutil
import sys
import tempfile
from contextlib import contextmanager
from pathlib import Path
from typing import Generator, List
from unittest import mock

import pytest

from snowflake.cli._plugins.workspace.context import WorkspaceContext, ActionContext
from snowflake.cli.api.console.abc import AbstractConsole

PROJECT_DIR = Path(__file__).parent / "test_data" / "projects"


@pytest.fixture
def temporary_directory():
    initial_dir = os.getcwd()

    with tempfile.TemporaryDirectory() as tmp_dir:
        try:
            os.chdir(tmp_dir)
            yield tmp_dir
        finally:
            os.chdir(initial_dir)


# Borrowed from tests_integration/test_utils.py
# TODO: remove from here when testing utils become shared
# TODO: contextlib.chdir isn't available before Python 3.11, so this is an alternative for older versions
@contextmanager
def change_directory(directory: Path):
    cwd = os.getcwd()
    os.chdir(directory)
    try:
        yield directory
    finally:
        os.chdir(cwd)


@pytest.fixture()
def workspace_context():
    return WorkspaceContext(
        console=mock.MagicMock(spec=AbstractConsole),
        project_root=Path().resolve(),
        get_default_role=lambda: "mock_role",
        get_default_warehouse=lambda: "mock_warehouse",
    )


@pytest.fixture()
def action_context():
    return ActionContext(get_entity=lambda *args: None)


@pytest.fixture
def snowflake_home(monkeypatch):
    """
    Set up the default location of config files to [temporary_directory]/.snowflake
    """
    with tempfile.TemporaryDirectory() as tmp_dir:
        snowflake_home = Path(tmp_dir) / ".snowflake"
        snowflake_home.mkdir()
        monkeypatch.setenv("SNOWFLAKE_HOME", str(snowflake_home))
        for module in [
            sys.modules["snowflake.connector.constants"],
            sys.modules["snowflake.connector.config_manager"],
            sys.modules["snowflake.connector.log_configuration"],
            sys.modules["snowflake.cli.api.config"],
        ]:
            importlib.reload(module)

        yield snowflake_home


@pytest.fixture
def alter_snowflake_yml():
    def _update(snowflake_yml_path: Path, parameter_path: str, value=None):
        import yaml

        with open(snowflake_yml_path) as fh:
            yml = yaml.safe_load(fh)

        parts = parameter_path.split(".")
        current_object = yml
        while parts:
            part = parts.pop(0)
            evaluated_part = int(part) if part.isdigit() else part

            if parts:
                if isinstance(current_object, dict):
                    current_object = current_object.setdefault(evaluated_part, {})
                else:
                    current_object = current_object[evaluated_part]
            else:
                current_object[evaluated_part] = value

        with open(snowflake_yml_path, "w+") as fh:
            yaml.safe_dump(yml, fh)

    return _update
