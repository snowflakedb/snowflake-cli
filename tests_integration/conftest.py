# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import functools
import json
import os
import shutil
import tempfile
from contextlib import contextmanager
from dataclasses import dataclass
from json import JSONDecodeError
from pathlib import Path
from typing import Any, Dict, List, Optional
from unittest import mock
from uuid import uuid4

import pytest
import yaml
from typer import Typer
from typer.testing import CliRunner

from snowflake.cli._app.cli_app import CliAppFactory
from snowflake.cli.api.cli_global_context import (
    fork_cli_context,
    get_cli_context_manager,
)
from snowflake.cli.api.connections import OpenConnectionCache
from snowflake.cli.api.project.util import TEST_RESOURCE_SUFFIX_VAR
from tests.conftest import clean_logging_handlers_fixture  # noqa: F401
from tests.testing_utils.files_and_dirs import merge_left
from tests.testing_utils.fixtures import (
    alter_snowflake_yml,  # noqa: F401
    snowflake_home,
)

pytest_plugins = [
    "tests_common",
    "tests_common.deflake",
    "tests_integration.testing_utils",
    "tests_integration.snowflake_connector",
]

TEST_DIR = Path(__file__).parent
DEFAULT_TEST_CONFIG = "connection_configs.toml"
WORLD_READABLE_CONFIG = "world_readable.toml"
IS_QA = "qa" in os.getenv("SNOWFLAKE_CONNECTIONS_INTEGRATION_HOST", "").lower()


@dataclass
class CommandResult:
    exit_code: int
    json: Optional[List[Dict[str, Any]] | Dict[str, Any]] = None
    output: Optional[str] = None
    stderr: Optional[str] = None


class TestConfigProvider:
    def __init__(self, temp_dir_with_configs: Path):
        self._temp_dir_with_configs = temp_dir_with_configs

    def get_config_path(self, file_name: str) -> Path:
        return self._temp_dir_with_configs / file_name


@pytest.fixture(scope="session")
def test_snowcli_config_provider():
    with tempfile.TemporaryDirectory() as td:
        temp_dst = Path(td) / "config"
        shutil.copytree(TEST_DIR / "config", temp_dst)
        for config_file in temp_dst.glob("**/*.toml"):
            if config_file.name != WORLD_READABLE_CONFIG:
                config_file.chmod(0o600)  # Make config file private
        yield TestConfigProvider(temp_dst)


@pytest.fixture(scope="session")
def test_root_path():
    return TEST_DIR


class SnowCLIRunner(CliRunner):
    def __init__(
        self,
        app: Typer,
        test_config_provider: TestConfigProvider,
        default_username: str,
        resource_suffix: str,
    ):
        super().__init__()
        self.app = app
        self._test_config_provider = test_config_provider
        self._test_config_path = self._test_config_provider.get_config_path(
            DEFAULT_TEST_CONFIG
        )
        self._default_username = default_username
        self._resource_suffix = resource_suffix

    def use_config(self, config_file_name: str) -> None:
        self._test_config_path = self._test_config_provider.get_config_path(
            config_file_name
        )

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        if "catch_exceptions" not in kw:
            kw.update(catch_exceptions=False)
        kw = self._with_env_vars(kw)

        # between every invocation, we need to reset the CLI context
        # and ensure no connections are cached going forward (to prevent
        # test cases from impacting each other / align with CLI usage)
        with fork_cli_context():
            connection_cache = OpenConnectionCache()
            cli_context_manager = get_cli_context_manager()
            cli_context_manager.reset()
            cli_context_manager.connection_cache = connection_cache
            try:
                return super().invoke(self.app, *a, **kw)
            finally:
                connection_cache.clear()

    def _with_env_vars(self, kw) -> dict:
        """
        Add required env vars to the invocation context if necessary and return new kwargs.

        Sets the USER env var to a default value if not set in the test,
        to allow us to use <% ctx.env.USER %> in test data on Windows.

        Sets the resource suffix env var unconditionally.
        The CLI automatically appends the value of this env var to some
        created resource identifiers, let's use this behaviour to add a unique
        suffix to resources used in tests to allow us to run simultaneous instances.
        """
        env = kw.get("env", {})
        return {
            **kw,
            "env": {
                **env,
                "USER": env.get("USER", self._default_username),
                TEST_RESOURCE_SUFFIX_VAR: self._resource_suffix,
            },
        }

    def invoke_with_config(self, args, **kwargs) -> CommandResult:
        result = self.invoke(
            [
                "--config-file",
                self._test_config_path,
                *args,
            ],
            **kwargs,
        )

        if result.output == "" or result.output.strip() == "Done":
            return CommandResult(result.exit_code, json=[])
        try:
            return CommandResult(
                result.exit_code, json.loads(result.output), output=result.output
            )
        except JSONDecodeError:
            return CommandResult(result.exit_code, output=result.output)

    def invoke_json(self, args, **kwargs) -> CommandResult:
        return self.invoke_with_config([*args, "--format", "JSON"], **kwargs)

    def invoke_with_connection_json(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_json([*args, "-c", connection], **kwargs)

    def invoke_with_connection(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_with_config([*args, "-c", connection], **kwargs)


@pytest.fixture
def runner(test_snowcli_config_provider, default_username, resource_suffix):
    app = CliAppFactory().create_or_get_app()
    yield SnowCLIRunner(
        app,
        test_snowcli_config_provider,
        default_username,
        resource_suffix,
    )


class QueryResultJsonEncoderError(RuntimeError):
    def __init__(self, output: str):
        super().__init__(f"Can not parse query result:\n{output}")


@pytest.fixture
def project_directory(temporary_working_directory, test_root_path):
    @contextmanager
    def _temporary_project_directory(
        project_name,
        merge_project_definition: Optional[dict] = None,
        subpath: Optional[Path] = None,
    ):
        test_data_file = test_root_path / "test_data" / "projects" / project_name
        project_dir = temporary_working_directory
        if subpath:
            project_dir = temporary_working_directory / subpath
            project_dir.mkdir(parents=True)
        shutil.copytree(test_data_file, project_dir, dirs_exist_ok=True)
        if merge_project_definition:
            with Path("snowflake.yml").open("r") as fh:
                project_definition = yaml.safe_load(fh)
            merge_left(project_definition, merge_project_definition)
            with open(Path(project_dir) / "snowflake.yml", "w") as file:
                yaml.dump(project_definition, file)

        yield project_dir

    return _temporary_project_directory


@pytest.fixture(autouse=True)
def reset_global_context_after_each_test(request):
    get_cli_context_manager().reset()
    yield


# This automatically used fixture isolates default location
# of config files from user's system.
@pytest.fixture(autouse=True)
def isolate_snowflake_home(snowflake_home):
    yield snowflake_home


@pytest.fixture
def default_username():
    return "snowflake"


@pytest.fixture
def resource_suffix(request):
    """
    Generate a random identifier suffix that includes the current test name.

    This suffix will be added to certain created resources like Native App
    packages and applications to be able to detect tests that don't
    clean up properly. The UUID is to avoid conflicts between concurrent runs.
    """
    # To generate a suffix that isn't too long or complex, we use originalname, which is the
    # "bare" test function name, without filename, class name, or parameterization variables
    return f"_{uuid4().hex}_{request.node.originalname}"


@pytest.fixture
def enable_snowpark_glob_support_feature_flag():
    with mock.patch(
        f"snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_enabled",
        return_value=True,
    ), mock.patch(
        f"snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_disabled",
        return_value=False,
    ):
        yield
