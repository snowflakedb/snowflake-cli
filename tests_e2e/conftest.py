# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import shutil
import subprocess
import sys
from contextlib import contextmanager
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import Optional

import pytest
from snowflake.cli import __about__
from snowflake.cli.api.secure_path import SecurePath

from tests_common import IS_WINDOWS

TEST_DIR = Path(__file__).parent

pytest_plugins = [
    "tests_common",
]


def _clean_output(text: str):
    """
    Replacing util to clean up console output. Typer is using rich.Panel to show the --help content.
    Unfortunately Typer implementation hardcodes box style for Panel.
    The typer.rich_utils.STYLE_OPTIONS_TABLE_BOX works only for content within the Panel.
    """
    if text is None:
        return None
    return (
        text.replace("│", "|")
        .replace("─", "-")
        .replace("╭", "+")
        .replace("╰", "+")
        .replace("╯", "+")
        .replace("╮", "+")
    )


def subprocess_check_output(cmd, stdin: Optional[str] = None):
    try:
        output = subprocess.check_output(
            cmd, input=stdin, shell=IS_WINDOWS, stderr=sys.stdout, encoding="utf-8"
        )
        return _clean_output(output)
    except subprocess.CalledProcessError as err:
        print(err.output)
        raise


def subprocess_run(cmd, stdin: Optional[str] = None):
    p = subprocess.run(
        cmd,
        input=stdin,
        shell=IS_WINDOWS,
        capture_output=True,
        text=True,
        encoding="utf-8",
    )
    p.stdout = _clean_output(p.stdout)
    p.stderr = _clean_output(p.stderr)
    return p


@pytest.fixture(scope="session")
def test_root_path():
    return TEST_DIR


@pytest.fixture(autouse=True)
def disable_colors_and_styles_in_output(monkeypatch):
    """
    Colors and styles in output cause mismatches in asserts,
    this environment variable turn off styling
    """
    monkeypatch.setenv("TERM", "unknown")


@pytest.fixture(scope="session")
def snowcli(test_root_path):
    with TemporaryDirectory() as tmp_dir:
        tmp_dir_path = Path(tmp_dir)
        _create_venv(tmp_dir_path)
        _build_snowcli(tmp_dir_path, test_root_path)
        _install_snowcli_with_external_plugin(tmp_dir_path, test_root_path)
        if IS_WINDOWS:
            yield tmp_dir_path / "Scripts" / "snow.exe"
        else:
            yield tmp_dir_path / "bin" / "snow"


@pytest.fixture(autouse=True)
def isolate_default_config_location(monkeypatch, temp_dir):
    monkeypatch.setenv("SNOWFLAKE_HOME", temp_dir)


def _create_venv(tmp_dir: Path) -> None:
    subprocess_check_output(["python", "-m", "venv", tmp_dir])


def _build_snowcli(venv_path: Path, test_root_path: Path) -> None:
    subprocess_check_output(
        [_python_path(venv_path), "-m", "pip", "install", "--upgrade", "build"],
    )
    subprocess_check_output(
        [_python_path(venv_path), "-m", "build", test_root_path / ".."]
    )


def _pip_install(python, *args):
    return subprocess_check_output([python, "-m", "pip", "install", *args])


def _install_snowcli_with_external_plugin(
    venv_path: Path, test_root_path: Path
) -> None:
    version = __about__.VERSION
    python = _python_path(venv_path)
    _pip_install(
        python,
        test_root_path / f"../dist/snowflake_cli-{version}-py3-none-any.whl",
    )
    _pip_install(
        python,
        test_root_path.parent
        / "test_external_plugins"
        / "multilingual_hello_command_group",
    )

    # Required by snowpark example tests
    _pip_install(python, "snowflake-snowpark-python[pandas]==1.25.0")


def _python_path(venv_path: Path) -> Path:
    if IS_WINDOWS:
        return venv_path / "Scripts" / "python.exe"
    return venv_path / "bin" / "python"


# Inspired by project_directory fixture in tests_integration/conftest.py
# This is a simpler implementation of that fixture, i.e. does not include supporting local PDFs.
@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory


@pytest.fixture()
def prepare_test_config_file(temp_dir):
    def f(config_file_path: SecurePath):
        target_file_path = Path(temp_dir) / "config.toml"
        config_file_path.copy(target_file_path)
        return target_file_path

    return f


@pytest.fixture()
def config_file(test_root_path, prepare_test_config_file):
    yield prepare_test_config_file(
        SecurePath(test_root_path) / "config" / "config.toml"
    )


@pytest.fixture()
def empty_config_file(test_root_path, prepare_test_config_file):
    yield prepare_test_config_file(SecurePath(test_root_path) / "config" / "empty.toml")


@pytest.fixture()
def example_connection_config_file(test_root_path, prepare_test_config_file):
    yield prepare_test_config_file(
        SecurePath(test_root_path) / "config" / "example_connection.toml"
    )
