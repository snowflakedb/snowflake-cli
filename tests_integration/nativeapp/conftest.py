from __future__ import annotations

from contextlib import contextmanager
from pathlib import Path
from typing import Any

import pytest
import yaml

from tests.nativeapp.factories import (
    ApplicationEntityModelFactory,
    ApplicationPackageEntityModelFactory,
    ManifestFactory,
    ProjectV2Factory,
)
from tests_integration.conftest import SnowCLIRunner


@pytest.fixture
def nativeapp_project_directory(project_directory, nativeapp_teardown):
    """Wrapper around the project_directory fixture specific to Native App testing.

    This fixture provides a context manager that does the following:
    - Automatically calls `snow app teardown` before exiting

    Parameters for the returned context manager:
    :param name: The name of the directory in tests_integration/test_data/projects to use.
    """

    @contextmanager
    def _nativeapp_project_directory(name, teardown_args: list[str] | None = None):
        with project_directory(name) as d:
            with nativeapp_teardown(project_dir=d, extra_args=teardown_args):
                yield d

    return _nativeapp_project_directory


@pytest.fixture
def nativeapp_teardown(runner: SnowCLIRunner):
    """Runs `snow app teardown` before exiting.

    This fixture provides a context manager that runs
    `snow app teardown --force --cascade` before exiting,
    regardless of any exceptions raised.

    Parameters for the returned context manager:
    :param project_dir: Path to the project directory (optional)
    :param env: Environment variables to replace os.environ (optional)
    """

    @contextmanager
    def _nativeapp_teardown(
        *,
        project_dir: Path | None = None,
        env: dict | None = None,
        extra_args: list[str] | None = None,
    ):
        # Back up project definition files in case the test
        # modifies them and makes the teardown fail
        backups = {
            path: path.read_text()
            for p in ["snowflake.yml", "snowflake.local.yml"]
            if (path := (project_dir or Path()) / p).exists()
        }
        try:
            yield
        finally:
            for path, backup in backups.items():
                path.write_text(backup)

            args = ["--force", "--cascade"]
            if project_dir:
                args += ["--project", str(project_dir)]
            if extra_args:
                args += extra_args
            kwargs: dict[str, Any] = {}
            if env:
                kwargs["env"] = env

            # `snow app teardown` can only teardown one package at a time for safety,
            # so when cleaning up PDFv2 tests, we need to iterate all the package entities
            # and teardown each one individually.
            snowflake_yml = (project_dir or Path.cwd()) / "snowflake.yml"
            with open(snowflake_yml, "r") as f:
                project_yml = yaml.safe_load(f)
            packages = [
                entity_id
                for entity_id, entity in project_yml.get("entities", {}).items()
                if entity["type"] == "application package"
            ]
            if packages:
                for package in packages:
                    result = runner.invoke_with_connection(
                        ["app", "teardown", *args, "--package-entity-id", package],
                        **kwargs,
                    )
                    assert result.exit_code == 0
            else:
                result = runner.invoke_with_connection(
                    ["app", "teardown", *args], **kwargs
                )
                assert result.exit_code == 0

    return _nativeapp_teardown


@pytest.fixture
def nativeapp_basic_pdf(temp_dir):
    return ProjectV2Factory(
        pdf__entities=dict(
            pkg=ApplicationPackageEntityModelFactory(
                identifier="my_pkg",
                artifacts=[{"src": "*", "dest": "./"}],
            ),
            app=ApplicationEntityModelFactory(
                identifier="myapp",
                fromm__target="pkg",
            ),
        ),
        files={
            "setup.sql": "CREATE OR ALTER VERSIONED SCHEMA core;",
            "README.md": "\n",
            "manifest.yml": ManifestFactory(),
        },
    )
