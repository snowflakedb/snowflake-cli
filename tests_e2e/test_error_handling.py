# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
from pathlib import Path
from textwrap import dedent

import pytest
from snowflake.cli.api.secure_utils import restrict_file_permissions

from tests_e2e.conftest import subprocess_run


@pytest.mark.e2e
def test_error_traceback_disabled_without_debug(snowcli, test_root_path, config_file):
    traceback_msg = "Traceback (most recent call last)"
    result = subprocess_run(
        [
            snowcli,
            "--config-file",
            config_file,
            "sql",
            "-c",
            "integration",
            "-q",
            "select foo",
        ]
    )

    assert "SQL compilation error" in result.stderr
    assert traceback_msg not in result.stderr
    assert result.stdout == "select foo\n"


@pytest.mark.e2e
def test_corrupted_config_in_default_location(
    snowcli,
    temporary_directory,
    isolate_default_config_location,
    test_root_path,
    snapshot,
):
    default_config = Path(temporary_directory) / "config.toml"
    default_config.write_text("[connections.demo]\n[connections.demo]")
    restrict_file_permissions(default_config)
    # corrupted config should produce human-friendly error
    result_err = subprocess_run(
        [snowcli, "connection", "list"],
    )
    assert result_err.stderr == snapshot

    # corrupted config in default location should not influence one passed with --config-file flag
    healthy_config = test_root_path / "config" / "config.toml"
    result_healthy = subprocess_run(
        [snowcli, "--config-file", healthy_config, "connection", "list"],
    )
    assert "dev" in result_healthy.stdout and "integration" in result_healthy.stdout


@pytest.mark.e2e
def test_initial_log_with_loaded_external_plugins_in_custom_log_path(
    snowcli, temporary_directory, isolate_default_config_location
):
    custom_log_path = os.path.join(temporary_directory, "custom", "logs")
    default_config = Path(temporary_directory) / "config.toml"
    config_logs_path = custom_log_path.replace("\\", "\\\\")
    default_config.write_text(
        dedent(
            f"""[cli.logs]
            path = "{config_logs_path}"

            [connections.default]
            [connections.integration]

            [cli.plugins.snowpark-hello]
            enabled = true
            [cli.plugins.snowpark-hello.config]
            greeting = "Hello"

            [cli.plugins.multilingual-hello]
            enabled = true
            """
        )
    )
    restrict_file_permissions(default_config)

    result = subprocess_run(
        [snowcli, "--help"],
    )

    assert result.returncode == 0
    with open(os.path.join(custom_log_path, "snowflake-cli.log")) as log_file:
        log_content = log_file.read()
        assert "Loaded external plugin: multilingual-hello" in log_content


@pytest.mark.e2e
def test_initial_log_with_loaded_external_plugins_in_custom_log_path_with_custom_config(
    snowcli, temporary_directory, isolate_default_config_location
):
    custom_log_path = os.path.join(temporary_directory, "custom", "logs")
    custom_config = Path(temporary_directory) / "custom" / "config.toml"
    custom_config.parent.mkdir()
    config_logs_path = custom_log_path.replace("\\", "\\\\")
    custom_config.write_text(
        dedent(
            f"""[cli.logs]
            path = "{config_logs_path}"

            [connections.default]
            [connections.integration]

            [cli.plugins.snowpark-hello]
            enabled = true
            [cli.plugins.snowpark-hello.config]
            greeting = "Hello"

            [cli.plugins.multilingual-hello]
            enabled = true
            """
        )
    )
    restrict_file_permissions(custom_config)

    result = subprocess_run([snowcli, "--config-file", custom_config, "--help"])

    assert result.returncode == 0
    with open(os.path.join(custom_log_path, "snowflake-cli.log")) as log_file:
        log_content = log_file.read()
        assert "Loaded external plugin: multilingual-hello" in log_content
