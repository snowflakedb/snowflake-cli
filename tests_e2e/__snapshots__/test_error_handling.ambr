# serializer version: 1
# name: test_corrupted_config_in_default_location
  '''
  +- Error ----------------------------------------------------------------------+
  | Configuration file seems to be corrupted. Key "demo" already exists. at line |
  | 2 col 18                                                                     |
  +------------------------------------------------------------------------------+
  '''
# ---
